python/ray/_private/function_manager.py

```diff
@@ -127,6 +132,25 @@ def load_function_or_class_from_local(self, module_name,
         except Exception:
             return None
 
+    def export_key(self, key):
+        """Export a key so it can be imported by other workers"""
+
+        # It's going to check all the keys until if reserve one key not
+        # existing in the cluster.
+        # One optimization is that we can use importer counter since
```

Comment:
Does the driver have an import thread? I'm concerned that the driver won't start pos at num_imported, and it will take a long time to probe the GCS (one check for each key).

Say there's 10k functions exported (this isn't unusual if done in a loop). Then a new driver needs to do 10k GCS lookups before starting.

Alternatively, we could do quadratic probing to check for a new key slot?

Reply:
```
1495 ┊   # Start the import thread
1496 ┊   if mode not in (RESTORE_WORKER_MODE, SPILL_WORKER_MODE):
1497 ┊   ┊   worker.import_thread = import_thread.ImportThread(
1498 ┊   ┊   ┊   worker, mode, worker.threads_stopped)
1499 ┊   ┊   worker.import_thread.start()
```
I think driver also has import thread? Only RESTORE_WORKER_MODE, SPILL_WORKER_MODE doesn't have import thread I think..

Reply:
Btw, quadratic probing is not going to work because we are doing check and set. to use quadratic we need to do: find and check-and-set. it might not be worth changing to that given we have num_imported here. How do you think?

Reply:
Hmm what if the driver or worker hasn't finished importing those yet and sees num_imported == 0? Then it might take a long time to scan from 0...10k and cause a lot of load on the GCS. Can we test this isn't an issue in practice?

Would it also work to use a separate counter for each job?

Reply:
> Would it also work to use a separate counter for each job?
It might make sense, but I don't want to change the logic here to this. It could cause corner cases maybe.
How about this, I introduce a counter in function_manager and next time when export, I use the max(export_counter, import_counter), and this should be good since it's O(N) then.